### DISCUSSION TITLE
Does xpra fits my need to have a framebuffer which works together with my nvidia GPU configured as PRIME RENDER OFFLOAD.

### DISCUSSION THREAD

**Discussion by Marietto2008** [Xpra-org/xpra#3720] [General]
Hello xpra users and developers,

I have two monitors right now. The first one is new,AOC 32 inches and I use it for FreeBSD. The second screen is 25 years old,it’s a SAMSUNG SYNCMASTER that I don’t want to use because it's not working well. I have 3 GPUS on my PC :

1) nvidia geforce GTX 1060 used for FreeBSD and configured as default on the BIOS / my PC boots using it /

2) Intel GPU integrated with the mobo

3) nvidia RTX 2080 ti. I want to use only the AOC monitor to use FreeBSD.

When I boot any bhyve (a hypervisor for BSD) / linux Vm I don’t want to use the Samsung monitor. I want to be able to use the desktop UI in Linux within a window with a framebuffer and at the same time I want to render my projects using the RTX 2080 ti in Linux,using the technology PRIME RENDER OFFLOAD,so without attaching any monitor to the VM. I’ve already installed the CUDA toolkit inside the Linux VM and it works well,BUT only if the RTX 2080 ti is attached to a physical monitor. At the moment the configuration that I’m trying to achieve is more comfortable for me than using two monitors. I need the framebuffer because without it,I can’t use Linux at all,since I can’t watch what happens when I move my mouse and keyboard,that’s because I can use one only monitor. The Samsung monitor is not working well. I think it’s dying. So I want :

    1. use one monitor connected to the first nvidia gpu (geforce 1060) and this monitor should display what happens on the host and even on the guest os (within the Linux desktop environment) and at the same time I would like to use the 2080 ti for rendering porpoises

    2. bound the 1060 to the host bsd : yes

    3. displaying the host’s graphical desktop and the graphical desktop of the linux - bhyve virtual machines at the same time,running the latter within a window smaller than the host screen resolution

    4. running a linux vm : yes,using bhyve

    5. with a passed-through second nvidia : yes,with the 2080 ti used for rendering

    6. with its output redirected to the host’s display : yes. I think that this can be done with a framebuffer. I always use the framebuffer created with vnc,but if at the same time I make the pass thru of the rtx 2080 ti I’m forced to use a secondary monitor,but I don’t want to do this.

    I’m not sure if the intel gpu can act as a framebuffer instead of the fbdev driver used by vnc or if there is a way to use a framebuffer (fbdev or any other,I don’t care) together with the 2080 ti,WITHOUT using a secondary monitor.

    virtualgl can allow me to display what happens inside the linux vm and at the same time can I render my 3D projects using the 2080 ti ? I want to enjoy myself with stable diffusion. The os I'm running is FreeBSD,but I can also use Linux,virtualizing it with the bhyve hypervisor. Generix says that VirtualGL works only with OpenGL but I don't know if stable diffusion can use the OpenGL acceleration...

I've created a thread on the nvidia forum,that you can read entirely going here :

https://forums.developer.nvidia.com/t/im-failing-trying-to-configure-the-framebuffer-as-default-graphic-adapter-and-the-nvidia-geforce-rtx-2080-ti-as-prime-render-offload/236365/37

The verdict of the nvidia developer has been something like this :

This won’t work with just one monitor, the limiting factor will always be the missing prime functions of the virtual gpu’s drm driver. I think it’s more fruitful to check if someone has some patches for it in the works.

You can only use VirtualGL or rather bumblebee. This needs to be installed on the Linux VM. It will spawn a second Xserver on the nVidia gpu (that you won’t see) and then copy the output to the Xserver on the virtual graphics. Downsides: limited to OpenGL, performance loss. 

At this point I've thought about Xpra. More or less I've understood what it is. But I'm not sure if it can solve my issue,I don't know if it allows me to do what I want. I suppose that what I need is a framebuffer that has the 3d acceleration with a low level of latencies that I can use within the Linux VM and the host os,which can act as an extension of the Linux graphical environment and that connects it to the graphical environment that I run on FreeBSD and that allows me to render my 3d projects using my RTX 2080 ti as PRIME RENDER OFFLOAD ? Is xpra exactly something like that or am I missing something ? thanks.

**Comment by totaam** [comment]
Cut & pasted (without linking back to): https://github.com/Xpra-org/xpra/issues/3718

Please see the suggestions in https://github.com/Xpra-org/xpra/issues/3718#issuecomment-1364675270

**Comment by totaam** [comment]
> Generix says that VirtualGL works only with OpenGL

Who is Generix?
> but I don't know if stable diffusion can use the OpenGL acceleration

Have you tried running stable diffusion via vgl? It should be trivial to try.
You seem to be posting the same question in lots of different places - which is frowned upon:
https://askubuntu.com/questions/1447224/can-bumblebee-offers-a-framebuffer-driver-display-server-that-can-works-togeth

> My main gpu is RTX 2080 ti,not GTX 2xxx
Same thing. `2xxx` means any `2***` GPU.

> The user case here is which kind of framebuffer is supported if I want to use my RTX 2080 ti gpu for rendering and at the same time I want to use the framebuffer to see what happens inside the VM based on Linux using one only monitor

I'm having trouble understanding your sentence.
> "framebuffer is supported"

doesn't explain what you want to do but focuses on a technical implementation detail, making it hard to understand and answer.
> and at the same time I want to use the framebuffer to see

Same here.
> based on Linux using one only monitor

I fail to see how the number of monitors is relevant.

**Reply by Marietto2008** [reply]
> > Generix says that VirtualGL works only with OpenGL
> 
> Who is Generix?

The nvidia developer who helped me on the nvidia forum,here : https://forums.developer.nvidia.com/t/im-failing-trying-to-configure-the-framebuffer-as-default-graphic-adapter-and-the-nvidia-geforce-rtx-2080-ti-as-prime-render-offload/236365/37

> > but I don't know if stable diffusion can use the OpenGL acceleration
> 
> Have you tried running stable diffusion via vgl? It should be trivial to try. You seem to be posting the same question in lots of different places - which is frowned upon: https://askubuntu.com/questions/1447224/can-bumblebee-offers-a-framebuffer-driver-display-server-that-can-works-togeth

I know. Since I'm not an experienced system administrator,I need to gather more informations from a lot of places to be able to understand what's the road to take
 
> > My main gpu is RTX 2080 ti,not GTX 2xxx
> > Same thing. `2xxx` means any `2***` GPU.
> 
> > The user case here is which kind of framebuffer is supported if I want to use my RTX 2080 ti gpu for rendering and at the same time I want to use the framebuffer to see what happens inside the VM based on Linux using one only monitor
> 
> I'm having trouble understanding your sentence.
> 
> > "framebuffer is supported"
> 
> doesn't explain what you want to do but focuses on a technical implementation detail, making it hard to understand and answer.
> 
> > and at the same time I want to use the framebuffer to see
> 
> Same here.
> 
> > based on Linux using one only monitor
> 
> I fail to see how the number of monitors is relevant.

this is the use case :

```
bhyve -S -c sockets=1,cores=2,threads=2 -m 4G -w -H -A \
-s 0,hostbridge \
-s 2,virtio-blk,/mnt/$vmdisk1'p2'/bhyve/img/Linux/ubuntu2210.img,bootindex=1 \
-s 6,fbuf,tcp=0.0.0.0:5919,w=1600,h=950,wait \
-s 8:0,passthru,2/0/0 \
-s 8:1,passthru,2/0/1 \
-s 8:2,passthru,2/0/2 \
-s 8:3,passthru,2/0/3 \
-s 10,virtio-net,tap19 \
-s 11,virtio-9p,sharename=/ \
-s 30,xhci,tablet \
-s 31,lpc \
-l bootrom,/usr/local/share/uefi-firmware/BHYVE_UEFI_CODE.fd \
vm0:19 < /dev/null & sleep 2 && vncviewer 0:19

```
These address are the PCI address of the RTX 2080 ti 

lspci :

02:00.0 VGA compatible controller: NVIDIA Corporation TU102 [GeForce RTX 2080 Ti] (rev a1)
02:00.1 Audio device: NVIDIA Corporation TU102 High Definition Audio Controller (rev a1)
02:00.2 USB controller: NVIDIA Corporation TU102 USB 3.1 Host Controller (rev a1)
02:00.3 Serial bus controller: NVIDIA Corporation TU102 USB Type-C UCSI Controller (rev a1)

I have used the same example on the post that I have created on the nvidia forum and I gave you the link at the beginning. Did you open that link ?

https://forums.developer.nvidia.com/t/im-failing-trying-to-configure-the-framebuffer-as-default-graphic-adapter-and-the-nvidia-geforce-rtx-2080-ti-as-prime-render-offload/236365/37

everything is pretty well explained there. Anyway,if you run bhyve (the most functional hypervisor for BSD) with the parameters that I have shown above,you will see that if you DON'T attach a monitor to the RTX 2080 ti,you will see nothing within the window framebuffer created by VNC. That's normal if you have two monitors. But at the moment I have one only monitor and I will be happy if I could see the Linux graphical environment inside the window framebuffer and at the same time,I would like to use my RTX 2080 ti with the PRIME RENDER OFFLOAD,using ONE only monitor. This is the technical explanations given by generix :

![Screenshot_2022-12-26_23-20-14](https://user-images.githubusercontent.com/755339/209586209-ac019a78-55ab-4ec0-9592-36b443c97d92.png)

So,I'm trying to understand which method to use to have a working framebuffer that allows me to use one only monitor when I virtualize Linux with bhyve and I want to use my RTX 2080 ti as prime render offload. VirtualGL ? Xwayland ? Xpra ? I don't know.

**Comment by totaam** [comment]
> this is the user case :

(followed by a wall of commands)
This is not a use case.
